File: expressjs/src/multi-app/scheduler.ts
import type { AppManager } from "./manager.js";
import type { InstrumentationConfig } from "../config/index.js";
import { processWorkflowTimeouts } from "../engine/workflow-scheduler.js";
import { processWebhookRetries } from "../engine/webhook-scheduler.js";
import { cleanupOldEvents } from "../instrument/cleanup.js";

// MultiAppScheduler runs workflow timeouts, webhook retries, and event cleanup across all apps.
export class MultiAppScheduler {
  private manager: AppManager;
  private instrConfig: InstrumentationConfig;
  private workflowTimer: ReturnType<typeof setInterval> | null = null;
  private webhookTimer: ReturnType<typeof setInterval> | null = null;
  private eventCleanupTimer: ReturnType<typeof setInterval> | null = null;

  constructor(manager: AppManager, instrConfig: InstrumentationConfig) {
    this.manager = manager;
    this.instrConfig = instrConfig;
  }

  start(): void {
    this.workflowTimer = setInterval(() => {
      this.processAllWorkflowTimeouts().catch((err) => {
        console.error("ERROR: multi-app workflow scheduler:", err);
      });
    }, 60_000);

    this.webhookTimer = setInterval(() => {
      this.processAllWebhookRetries().catch((err) => {
        console.error("ERROR: multi-app webhook scheduler:", err);
      });
    }, 30_000);

    // Event cleanup runs hourly
    if (this.instrConfig.enabled) {
      this.eventCleanupTimer = setInterval(() => {
        this.processAllEventCleanup().catch((err) => {
          console.error("ERROR: multi-app event cleanup:", err);
        });
      }, 3_600_000);
    }

    console.log("Multi-app scheduler started (workflows: 60s, webhooks: 30s, event cleanup: 1h)");
  }

  stop(): void {
    if (this.workflowTimer) {
      clearInterval(this.workflowTimer);
      this.workflowTimer = null;
    }
    if (this.webhookTimer) {
      clearInterval(this.webhookTimer);
      this.webhookTimer = null;
    }
    if (this.eventCleanupTimer) {
      clearInterval(this.eventCleanupTimer);
      this.eventCleanupTimer = null;
    }
  }

  private async processAllWorkflowTimeouts(): Promise<void> {
    for (const ac of this.manager.allContexts()) {
      try {
        await processWorkflowTimeouts(ac.store, ac.registry);
      } catch (err) {
        console.error(`ERROR: workflow timeouts for app ${ac.name}:`, err);
      }
    }
  }

  private async processAllWebhookRetries(): Promise<void> {
    for (const ac of this.manager.allContexts()) {
      try {
        await processWebhookRetries(ac.store);
      } catch (err) {
        console.error(`ERROR: webhook retries for app ${ac.name}:`, err);
      }
    }
  }

  private async processAllEventCleanup(): Promise<void> {
    for (const ac of this.manager.allContexts()) {
      try {
        const deleted = await cleanupOldEvents(ac.store.pool, this.instrConfig.retention_days);
        if (deleted > 0) {
          console.log(`Event cleanup for app ${ac.name}: deleted ${deleted} old events`);
        }
      } catch (err) {
        console.error(`ERROR: event cleanup for app ${ac.name}:`, err);
      }
    }
  }
}

File: backend/elixir-phoenix/lib/rocket/multiapp/scheduler.ex
defmodule Rocket.MultiApp.Scheduler do
  @moduledoc "Multi-app background scheduler: iterates all app contexts for workflow timeouts and webhook retries."
  use GenServer
  require Logger

  alias Rocket.MultiApp.AppManager

  @workflow_interval 60_000
  @webhook_interval 30_000
  @cleanup_interval 3_600_000

  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
  end

  @impl true
  def init(_opts) do
    schedule_workflows()
    schedule_webhooks()
    schedule_cleanup()
    {:ok, %{}}
  end

  @impl true
  def handle_info(:workflow_tick, state) do
    process_all_workflow_timeouts()
    schedule_workflows()
    {:noreply, state}
  end

  def handle_info(:webhook_tick, state) do
    process_all_webhook_retries()
    schedule_webhooks()
    {:noreply, state}
  end

  def handle_info(:cleanup_tick, state) do
    process_all_event_cleanup()
    schedule_cleanup()
    {:noreply, state}
  end

  def handle_info(_msg, state), do: {:noreply, state}

  defp schedule_workflows do
    Process.send_after(self(), :workflow_tick, @workflow_interval)
  end

  defp schedule_webhooks do
    Process.send_after(self(), :webhook_tick, @webhook_interval)
  end

  defp schedule_cleanup do
    Process.send_after(self(), :cleanup_tick, @cleanup_interval)
  end

  defp process_all_workflow_timeouts do
    contexts = AppManager.all_contexts()

    Enum.each(contexts, fn ctx ->
      try do
        Rocket.Engine.WorkflowEngine.process_timeouts(ctx.db_pool, ctx.registry)
      rescue
        e -> Logger.error("Workflow timeout error for app #{ctx.name}: #{inspect(e)}")
      end
    end)
  rescue
    e -> Logger.error("Multi-app workflow scheduler error: #{inspect(e)}")
  end

  defp process_all_webhook_retries do
    contexts = AppManager.all_contexts()

    Enum.each(contexts, fn ctx ->
      try do
        Rocket.Engine.WebhookScheduler.process_retries_for(ctx.db_pool)
      rescue
        e -> Logger.error("Webhook retry error for app #{ctx.name}: #{inspect(e)}")
      end
    end)
  rescue
    e -> Logger.error("Multi-app webhook scheduler error: #{inspect(e)}")
  end

  defp process_all_event_cleanup do
    instr_config = Application.get_env(:rocket, :instrumentation_config) || %{enabled: true, retention_days: 7}
    if instr_config.enabled do
      contexts = AppManager.all_contexts()
      Enum.each(contexts, fn ctx ->
        try do
          Rocket.Instrument.Cleanup.cleanup_old_events(ctx.db_pool, instr_config.retention_days)
        rescue
          e -> Logger.error("Event cleanup error for app #{ctx.name}: #{inspect(e)}")
        end
      end)
    end
  rescue
    e -> Logger.error("Multi-app event cleanup error: #{inspect(e)}")
  end
end
